File: lib/normalize.rb
require "securerandom"
require "shellwords"
require_relative "cloud"
require_relative "cloud_move"
require_relative "normalization_plan"

module Cloud
  class Normalize
    # Keep construction and execution together so CLI entry points do not duplicate the workflow.
    def self.call(project, bucket)
      new(project, bucket).call
    end

    # Keep the target fixed throughout the operation to avoid acting on a different bucket.
    def initialize(project, bucket)
      @project = project
      @bucket = bucket
    end

    # Validate names before staging objects so failures can restore the original layout.
    def call
      Cloud.login
      Cloud.exec(Shellwords.join(["gcloud", "config", "set", "project", @project]))
      # The API returns object names as data, so wildcard characters cannot alter the listing.
      files = Cloud.pipe(Cloud::StorageApi.list_command(@bucket), &:readlines).map(&:chomp)

      normalize_objects(NormalizationPlan.build(files).reject { |source, target| source == target })
    end

    private

    def normalize_objects(entries)
      moves = entries.map { |source, target| [source, target, temporary_path(source)] }
      staged = []
      finalized = []

      process_moves(moves, staged, finalized)
    end

    def process_moves(moves, staged, finalized)
      stage_objects(moves, staged)
      finalize_objects(moves, finalized)
    rescue StandardError, SignalException => e
      rollback_after_failure(e, staged, finalized)
    end

    def rollback_after_failure(error, staged, finalized)
      rollback_errors = rollback_objects(staged, finalized)
      NormalizationPlan.raise_on_rollback_failure(error, rollback_errors)
      raise error
    end

    def stage_objects(moves, staged)
      moves.each do |source, _target, temporary|
        record_move(source, temporary, staged)
      end
    end

    def finalize_objects(moves, finalized)
      moves.each do |source, target, temporary|
        record_move(temporary, target, finalized) { |move| move[0] = source }
      end
    end

    def record_move(source, target, moves)
      # Defer signals until the move and generation are recorded so rollback sees every side effect.
      Thread.handle_interrupt(SignalException => :never) do
        execute_and_record_move(source, target, moves) { |move| yield move if block_given? }
      end
    end

    def execute_and_record_move(source, target, moves)
      target_generation = move_object(source, target)
      moves << [source, target, target_generation]
      yield moves.last
    rescue Cloud::CommandError, Cloud::ObjectMove::MissingGenerationError, IOError, SystemCallError => e
      Cloud::ObjectMove.confirm_move_after_failure(source, target, e) unless target_generation
      raise
    end

    def temporary_path(source)
      "#{source}.task-googlecloud-#{SecureRandom.hex(16)}"
    end

    def move_object(source, target)
      target_generation = Cloud::ObjectMove.move(source, target)
      sleep 1
      target_generation
    end

    def rollback_objects(staged, finalized)
      rollback_finalized(finalized) + rollback_staged(staged, finalized)
    end

    def rollback_finalized(finalized)
      finalized.reverse_each.filter_map do |temporary, target, target_generation|
        attempt_rollback(temporary, target, target_generation)
      end
    end

    def rollback_staged(staged, finalized)
      finalized_sources = finalized.map(&:first)
      staged.reverse_each.filter_map do |source, temporary, temporary_generation|
        # Finalization already restored this source; replaying staging would use an obsolete generation.
        next if finalized_sources.include?(source)

        attempt_rollback(source, temporary, temporary_generation)
      end
    end

    # Try every known move so one cleanup failure does not block later restorations.
    def attempt_rollback(source, target, target_generation)
      Cloud.exec(Cloud::ObjectMove.rollback_command(source, target, target_generation))
    rescue StandardError => e
      e
    end
  end
end

File: src/normalize.rs
use crate::InterruptFlag;
use crate::cloud::Cloud;
use crate::error::AppError;
use crate::normalization_plan;
use crate::object_move;
use crate::storage::{MAX_OBJECT_NAME_BYTES, ObjectPath, StorageClient};
use crate::transaction::{RemoteChange, rollback_moves};

const TEMPORARY_SUFFIX_PREFIX: &str = ".task-googlecloud-";

pub fn run<S: StorageClient>(
    cloud: &Cloud,
    storage: &S,
    interrupt: &InterruptFlag,
    project: &str,
    bucket: &str,
) -> Result<(), AppError> {
    cloud.login()?;
    cloud.set_project(project)?;
    let files = storage.list_objects(bucket)?;
    let plan = normalization_plan::build(&files)?;
    let moves = plan_moves(plan)?;
    process_moves(storage, interrupt, moves)
}

fn plan_moves(
    plan: Vec<normalization_plan::Entry>,
) -> Result<Vec<(ObjectPath, ObjectPath, ObjectPath)>, AppError> {
    plan.into_iter()
        .filter(|entry| entry.source != entry.target)
        .map(|entry| {
            let source = ObjectPath::parse(&entry.source)?;
            let target = ObjectPath::parse(&entry.target)?;
            validate_target_path(&target)?;
            let temporary = temporary_path(&source)?;
            Ok((source, target, temporary))
        })
        .collect()
}

fn validate_target_path(target: &ObjectPath) -> Result<(), AppError> {
    if target.object.len() > MAX_OBJECT_NAME_BYTES {
        return Err(AppError::Message(format!(
            "Object name is too long for normalized target: {}",
            target.uri()
        )));
    }
    Ok(())
}

pub fn process_moves<S: StorageClient>(
    storage: &S,
    interrupt: &InterruptFlag,
    moves: Vec<(ObjectPath, ObjectPath, ObjectPath)>,
) -> Result<(), AppError> {
    let mut staged = Vec::new();
    let mut finalized = Vec::new();

    let operation = (|| {
        for (source, _target, temporary) in &moves {
            let generation = object_move::execute(storage, interrupt, source, temporary, None)?;
            staged.push(RemoteChange {
                source: source.clone(),
                target: temporary.clone(),
                generation,
            });
            interrupt.check()?;
        }

        for (change, (_, target, _)) in staged.iter().zip(&moves) {
            let generation = object_move::execute(
                storage,
                interrupt,
                &change.target,
                target,
                Some(&change.generation),
            )?;
            finalized.push(RemoteChange {
                source: change.source.clone(),
                target: target.clone(),
                generation,
            });
            interrupt.check()?;
        }
        Ok::<(), AppError>(())
    })();

    match operation {
        Ok(()) => Ok(()),
        Err(error) => {
            interrupt.clear_for_rollback();
            Err(AppError::rollback(
                error,
                rollback_moves(storage, &staged, &finalized),
            ))
        }
    }
}

fn temporary_path(source: &ObjectPath) -> Result<ObjectPath, AppError> {
    let suffix = temporary_suffix();
    if source.object.len() + suffix.len() > MAX_OBJECT_NAME_BYTES {
        return Err(AppError::Message(format!(
            "Object name is too long for temporary staging: {}",
            source.uri()
        )));
    }

    Ok(ObjectPath {
        bucket: source.bucket.clone(),
        object: format!("{}{}", source.object, suffix),
    })
}

fn temporary_suffix() -> String {
    format!("{TEMPORARY_SUFFIX_PREFIX}{}", uuid::Uuid::new_v4().simple())
}

#[cfg(test)]
#[path = "../tests/unit/normalize.rs"]
mod tests;
